sdk/go/common/workspace/templates_zip.go

```diff
@@ -0,0 +1,96 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package workspace
+
+import (
+	"archive/zip"
+	"bytes"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"os"
+	"strings"
+)
+
+func isZIPTemplateURL(templateNamePathOrURL string) bool {
+	parsedURL, _ := url.Parse(templateNamePathOrURL)
+	return parsedURL.Path != "" && strings.HasSuffix(parsedURL.Path, ".zip")
+}
+
+func retrieveZIPTemplates(templateURL string) (TemplateRepository, error) {
+	var err error
+	// Create a temp dir.
+	var temp string
+	if temp, err = os.MkdirTemp("", "pulumi-template-"); err != nil {
+		return TemplateRepository{}, err
+	}
+
+	parsedURL, err := url.Parse(strings.ReplaceAll(templateURL, ".zip", ""))
+	if err != nil {
+		return TemplateRepository{}, err
+	}
+
+	var fullPath string
+	if fullPath, err = RetrieveZIPTemplateFolder(parsedURL, temp); err != nil {
+		return TemplateRepository{}, fmt.Errorf("failed to retrieve zip archive: %w", err)
+	}
+
+	return TemplateRepository{
+		Root:         temp,
+		SubDirectory: fullPath,
+		ShouldDelete: true,
+	}, nil
+}
+
+func RetrieveZIPTemplateFolder(templateURL *url.URL, tempDir string) (string, error) {
+	packageRequest, err := http.NewRequest(http.MethodPut, templateURL.String(), bytes.NewReader([]byte{}))
+	if err != nil {
+		return "", err
+	}
+	packageRequest.Header.Set("Accept", "application/gzip")
+	requestQuery := packageRequest.URL.Query()
+	for key, values := range templateURL.Query() {
+		for _, value := range values {
+			requestQuery.Add(key, value)
+		}
+	}
+	packageRequest.URL.RawQuery = requestQuery.Encode()
+	packageResponse, err := http.DefaultClient.Do(packageRequest)
+	if err != nil {
+		return "", err
+	}
+	packageResponseBody, err := io.ReadAll(packageResponse.Body)
+	if err != nil {
+		return "", err
+	}
+	archive, err := zip.NewReader(bytes.NewReader(packageResponseBody), int64(len(packageResponseBody)))
+	if err != nil {
+		return "", err
+	}
+	for _, file := range archive.File {
+		filePath := fmt.Sprintf("%s/%s", tempDir, file.Name)
+		fileReader, err := file.Open()
+		if err != nil {
+			return "", err
+		}
+		fileContents, err := io.ReadAll(fileReader)
+		if err != nil {
+			return "", err
+		}
+		os.WriteFile(filePath, fileContents, 0o600)
```

Comment:
Not checking for err here. Also could do an io.Copy rather than ReadAll/WriteAll?